sql/catalog.go

```diff
@@ -124,3 +133,50 @@ func (d Databases) Table(dbName string, tableName string) (Table, error) {
 
 	return table, nil
 }
+
+// LockTable adds a lock for the given table and session client. It is assumed
+// the database is the current database in use.
+func (c *Catalog) LockTable(id uint32, table string) {
+	db := c.CurrentDatabase()
+	c.mu.Lock()
+	defer c.mu.Unlock()
+	if _, ok := c.locks[id]; !ok {
+		c.locks[id] = make(dbLocks)
+	}
+
+	if _, ok := c.locks[id][db]; !ok {
+		c.locks[id][db] = make(tableLocks)
+	}
+
+	c.locks[id][db][table] = struct{}{}
```

Comment:
I'm not sure if we plan (in the future) lock all DB or Session (not only tables, rows), but if yes then maybe flat structure, e.g. `c.locks["id.db.table"]` may work better, because you don't need to create nested structs. Moreover when you unlock you have to propagate this info to the higher levels, instead of delete one key.
But frankly it's up to you.

Reply:
The thing is, we use the elements of each separate step. c.locks[id] to get all the locks for a session and then c.locks[id][db] to get the tables. We can't do that with a single key